lib/route/cls/flower.c

```diff
@@ -598,6 +670,124 @@ int rtnl_flower_get_ip_dscp(struct rtnl_cls *cls, uint8_t *dscp, uint8_t *mask)
 	return 0;
 }
 
+/**
+ * Set IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_src(struct rtnl_cls *cls, int addr, int mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_src = nl_addr_build(AF_INET,
+					       (unsigned char *)&addr,
+					       sizeof(__be32));
+
+		f->cf_mask |= FLOWER_ATTR_IPV4_SRC;
+
+		if (mask) {
+			uint32_t m = 0xFFFFFFFF << (32 - mask);
+			f->cf_ipv4_src_mask = htonl(m);
+			f->cf_mask |= FLOWER_ATTR_IPV4_SRC_MASK;
+		}
+
+		return 0;
+	}
+
+	return -NLE_FAILURE;
+}
+
+/**
+ * Get IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_get_ipv4_src(struct rtnl_cls *cls, struct nl_addr *addr,
+			     uint32_t *mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data_peek(TC_CAST(cls))))
+		return -NLE_INVAL;
+
+	if (!(f->cf_mask & FLOWER_ATTR_IPV4_SRC))
+		return -NLE_MISSING_ATTR;
+
+	if (addr)
+		memcpy(addr, f->cf_ipv4_src, AF_INET);
+
+	*mask = f->cf_ipv4_src_mask;
+
+	return 0;
+}
+
+/**
+ * Set IPv4 destination address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 destination address
+ * @arg mask		mask for IPv4 destination address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_dst(struct rtnl_cls *cls, int addr, int mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_dst = nl_addr_build(AF_INET,
+					       (unsigned char *)&addr,
+					       sizeof(__be32));
+
+		f->cf_mask |= FLOWER_ATTR_IPV4_DST;
+
+		if (mask) {
+			uint32_t m = 0xFFFFFFFF << (32 - mask);
+			f->cf_ipv4_dst_mask = htonl(m);
+			f->cf_mask |= FLOWER_ATTR_IPV4_DST_MASK;
+		}
+
+		return 0;
+	}
+
+	return -NLE_FAILURE;
+}
+
+/**
+ * Get IPv4 destination address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 destination address
+ * @arg mask		mask for IPv4 destination address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_get_ipv4_dst(struct rtnl_cls *cls, struct nl_addr *addr,
+			     uint32_t *mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data_peek(TC_CAST(cls))))
+		return -NLE_INVAL;
+
+	if (!(f->cf_mask & FLOWER_ATTR_IPV4_DST))
+		return -NLE_MISSING_ATTR;
+
+	if (addr)
+		memcpy(addr, f->cf_ipv4_dst, AF_INET);
+
+	*mask = f->cf_ipv4_dst_mask;
```

Comment:
there is one getter for both the address and the mask (which is good).
But internally, you track this in two separate fields, which are even announced on netlink separately.

It's not clear, why we check above for `if (!(f->cf_mask & FLOWER_ATTR_IPV4_DST))`, but here we access the mask even if it's not set. It seems, if the mask is not set, then it should default to 0xFFFFFFFF. Currently it defaults to zero.